apps/emqx_message_transformation/src/emqx_message_transformation_schema.erl

```diff
@@ -0,0 +1,331 @@
+%% Internal exports
+%%------------------------------------------------------------------------------
+
+%%------------------------------------------------------------------------------
+%% Internal fns
+%%------------------------------------------------------------------------------
+
+mk(Type, Meta) -> hoconsc:mk(Type, Meta).
+ref(Name) -> hoconsc:ref(?MODULE, Name).
+
+payload_serde_member_selector(all_union_members) ->
+    payload_serde_refs();
+payload_serde_member_selector({value, V}) ->
+    payload_serde_refs(V).
+
+payload_serde_refs() ->
+    [
+        payload_serde_none,
+        payload_serde_json,
+        payload_serde_avro,
+        payload_serde_protobuf
+    ].
+payload_serde_refs(#{<<"type">> := Type} = V) when is_atom(Type) ->
+    payload_serde_refs(V#{<<"type">> := atom_to_binary(Type)});
+payload_serde_refs(#{<<"type">> := <<"none">>}) ->
+    [ref(payload_serde_none)];
+payload_serde_refs(#{<<"type">> := <<"json">>}) ->
+    [ref(payload_serde_json)];
+payload_serde_refs(#{<<"type">> := <<"avro">>}) ->
+    [ref(payload_serde_avro)];
+payload_serde_refs(#{<<"type">> := <<"protobuf">>}) ->
+    [ref(payload_serde_protobuf)];
+payload_serde_refs(_Value) ->
+    Expected = lists:join(
+        " | ",
+        [
+            Name
+         || T <- payload_serde_refs(),
+            "payload_serde_" ++ Name <- [atom_to_list(T)]
+        ]
+    ),
+    throw(#{
+        field_name => type,
+        expected => iolist_to_binary(Expected)
+    }).
+
+ensure_array(undefined, _) -> undefined;
+ensure_array(L, _) when is_list(L) -> L;
+ensure_array(B, _) -> [B].
+
+validate_unique_names(Transformations0) ->
+    Transformations = emqx_utils_maps:binary_key_map(Transformations0),
+    do_validate_unique_names(Transformations, #{}).
+
+do_validate_unique_names(_Transformations = [], _Acc) ->
+    ok;
+do_validate_unique_names([#{<<"name">> := Name} | _Rest], Acc) when is_map_key(Name, Acc) ->
+    {error, <<"duplicated name: ", Name/binary>>};
+do_validate_unique_names([#{<<"name">> := Name} | Rest], Acc) ->
+    do_validate_unique_names(Rest, Acc#{Name => true}).
+
+validate_unique_topics(Topics) ->
+    Grouped = maps:groups_from_list(
+        fun(T) -> T end,
+        Topics
+    ),
+    DuplicatedMap = maps:filter(
+        fun(_T, Ts) -> length(Ts) > 1 end,
+        Grouped
+    ),
+    case maps:keys(DuplicatedMap) of
+        [] ->
+            ok;
+        Duplicated ->
+            Msg = iolist_to_binary([
+                <<"duplicated topics: ">>,
+                lists:join(", ", Duplicated)
+            ]),
+            {error, Msg}
+    end.
+
+validate_operations([]) ->
+    {error, <<"at least one operation must be defined">>};
+validate_operations([_ | _]) ->
+    ok.
+
+compile_variform(Expression, #{make_serializable := true}) ->
+    case is_binary(Expression) of
+        true ->
+            Expression;
+        false ->
+            emqx_variform:decompile(Expression)
```

Comment:
That's a bit surprising :) `compile_variform` -> `emqx_variform:decompile`

Reply:
Yes, it is.  :joy_cat: 
I guess it's due to `make_serializable => true`.

Taken from here: https://github.com/emqx/emqx/blob/bb9fb10ec54400076f4d43643a6722af9a2e6018/apps/emqx/src/emqx_schema.erl#L1774